Userland/Libraries/LibGfx/BitmapSequence.cpp

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2024, Zachary Huang <zack466@gmail.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/AnonymousBuffer.h>
+#include <LibGfx/Bitmap.h>
+#include <LibGfx/BitmapSequence.h>
+#include <LibGfx/Size.h>
+#include <LibIPC/Decoder.h>
+#include <LibIPC/Encoder.h>
+#include <LibIPC/File.h>
+
+namespace Gfx {
+
+BitmapSequence::BitmapSequence(Vector<Optional<NonnullRefPtr<Bitmap>>> bitmaps)
+    : m_bitmaps(move(bitmaps))
+{
+}
+
+}
+
+namespace IPC {
+
+template<>
+ErrorOr<void> encode(Encoder& encoder, Gfx::BitmapSequence const& bitmap_sequence)
+{
+    Vector<Optional<NonnullRefPtr<Gfx::Bitmap>>> bitmaps = bitmap_sequence.bitmaps();
+
+    TRY(encoder.encode_size(bitmaps.size()));
+
+    size_t total_buffer_size = 0;
+
+    for (auto const& bitmap_option : bitmaps) {
+        TRY(encoder.encode(bitmap_option.has_value()));
+
+        if (bitmap_option.has_value()) {
+            auto bitmap = bitmap_option.value();
+            TRY(encoder.encode(static_cast<u32>(bitmap->format())));
+            TRY(encoder.encode(static_cast<u32>(bitmap->alpha_type())));
+            TRY(encoder.encode(bitmap->size_in_bytes()));
+            TRY(encoder.encode(bitmap->size()));
+
+            total_buffer_size += bitmap_option.value()->size_in_bytes();
+        }
+    }
```

Comment:
I would be a bit more comfortable with the bit wrangling we actually called encode() on a 

`Vector<Optional<BitmapMetadata>` for this header.

That would look something like this, where you construct a vector of metadata from the bitmaps (without copying them!) and encode that:

```
auto const& bitmaps = bitmap_sequence.bitmaps();
Vector<Optional<BitmapMetadata> metadata;
metadata.reserve(bitmaps.size());

for (auto const& bitmap : bitmaps) {
    Optional<BitmapMetadata> m;
    // fill in m
    metadata.append_unchecked(m);
}

TRY(encoder.encode(metadata))
```